Different protocols in the same Service definition with type=LoadBalancer (#1)
janosi/enhancements: keps/sig-network/20200103-mixed-protocol-lb.md

    - [Alibaba](#alibaba)
    - [AWS](#aws)
    - [Azure](#azure)
    - [GCE](#gce)
    - [IBM Cloud](#ibm-cloud)
    - [OpenStack](#openstack)
    - [Oracle Cloud](#oracle-cloud)
    - [Tencent Cloud](#tencent-cloud)
  - [Risks and Mitigations](#risks-and-mitigations)
- [Design Details](#design-details)
  - [Option Control Alternatives](#option-control-alternatives)
    - [Annotiation in the Service definition](#annotiation-in-the-service-definition)
    - [Merging Services in CPI](#merging-services-in-cpi)
  - [Test Plan](#test-plan)
  - [Graduation Criteria](#graduation-criteria)
    - [Examples](#examples)
      - [Alpha -&gt; Beta Graduation](#alpha---beta-graduation)
      - [Beta -&gt; GA Graduation](#beta---ga-graduation)
      - [Removing a deprecated flag](#removing-a-deprecated-flag)
  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
  - [Version Skew Strategy](#version-skew-strategy)
- [Implementation History](#implementation-history)
- [Drawbacks [optional]](#drawbacks-optional)
- [Alternatives [optional]](#alternatives-optional)
- [Infrastructure Needed [optional]](#infrastructure-needed-optional)
<!-- /toc -->

## Release Signoff Checklist

> **ACTION REQUIRED:** In order to merge code into a release, there must be an issue in [kubernetes/enhancements] referencing this KEP and targeting a release milestone **before [Enhancement Freeze](https://github.com/kubernetes/sig-release/tree/master/releases)

Reviewer: Be sure to remove any of the instructions parts that are part of the template.

of the targeted release**.

For enhancements that make changes to code or processes/procedures in core Kubernetes i.e., [kubernetes/kubernetes], we require the following Release Signoff checklist to be completed.

Check these off as they are completed for the Release Team to track. These checklist items _must_ be updated for the enhancement to be released.

> - [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)

Reviewer: Make sure to crosslink this issue under enhancements or create one if it doesn't exist.

Author: I created the issue and added the link here.

- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
- [ ] Graduation criteria is in place
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).

**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
[kubernetes/website]: https://github.com/kubernetes/website

## Summary

> The `Different protocols in the same Service definition with type=LoadBalancer` feature enables the creation of such LoadBalancer Services that have different port definitions with different protocols. 

Reviewer: ```suggestion
This feature enables the creation of aLoadBalancer Services that has different port definitions with different protocols. 
```

Reviewer: Move the part in single quotes to the name of the PR

Author: Fixed

## Motivation

> The ultimate goal of the feature is to support the use cases of the users when they want to expose their applications via a single IP address but different L4 protocols on a cloud provided load balancer. 

Reviewer: ```suggestion
The ultimate goal of this feature is to support users that want to expose their applications via a single IP address but different L4 protocols with a cloud provided load-balancer. 
```

> The following issue and PR shows the interest of the users in this feature:

Reviewer: ```suggestion
The following issue and PR shows considerable interest from users that would benefit from this feature:
```

https://github.com/kubernetes/kubernetes/issues/23880
https://github.com/kubernetes/kubernetes/pull/75831

### Goals

The goals of this KEP are:

> - analyze the impact of this feature in relation considering the load balancer implementation of cloud providers

Reviewer: ```suggestion
- To analyze the impact of this feature with regard to current implementations of cloud-provider load-balancers
```

> - define how the activation of this feature could be made configurable if the load balancer implementations of the cloud providers requires such optional behavior

Reviewer: ```suggestion
- define how the activation of this feature could be made configurable if certain cloud-provider load-balancer implementations do not want to provide this feature
```

Reviewer: It will be important to iterate the intention of this feature with regard to conformance testing. If this is a feature that we intend to be portable, then we should have a section that includes a timeline for GA and applicable PRs for conformance tests. 

We should also spell out the non-goals in relation to the context provided above. If this is considered "unportable" due to not enough cloud-providers seeking to implement it, then we need to mention this in the non-goals.

### Non-Goals

## Proposal

The first thing proposed here is to lift the hardcoded limitation in Kubernetes that currently rejects Service definitions with different protocols if their type is LoadBalancer. Kubernetes would not reject Service definitions like this from that point:
```yaml
apiVersion: v1
kind: Service
metadata:
  name: mixed-protocol
spec:
  type: LoadBalancer
  ports:
    - name: dns-udp
      port: 53
      protocol: UDP
    - name: dns-tcp
      port: 53
      protocol: TCP
  selector:
    app: my-dns-server
  ```

> Once that limit is removed those Service definitions will reach the Cloud Provider LB controller implementations. The logic of the particular Cloud Provider LB controller and of course the actual capabilities and architecture of the backing Cloud Load Balancer services determines how the actual exposure of the application really mnaifests. For this reason it is important to understand the capabilities of those backing services and to design this feature accordingly.

Reviewer: ```suggestion
Once that limit is removed those Service definitions will reach the Cloud Provider LB controller implementations. The logic of the particular Cloud Provider LB controller and of course the actual capabilities and architecture of the backing Cloud Load Balancer services determines how the actual exposure of the application really manifests. For this reason it is important to understand the capabilities of those backing services and to design this feature accordingly.
```

### User Stories [optional]

#### Story 1

> As a Kubernetes cluster user I want to expose my application that provides its services on different protocols on a single cloud provider load balancer IP. In order to achieve this I want to define different `ports` with different protocols in my Service definition with `type: LoadBalancer`

Reviewer: ```suggestion
As a Kubernetes cluster user I want to expose an application that provides its service on different protocols with a single cloud provider load balancer IP. In order to achieve this I want to define different `ports` with mixed protocols in my Service definition of `type: LoadBalancer`
```

### Implementation Details/Notes/Constraints [optional]

#### Alibaba

> The Alibaba CPI supports TCP, UDP and HTTPS in Service definitions and can configure the SLB listeners with the protocols defined in the Service.

Reviewer: Be sure to introduce the acronym `CPI` before using it in these examples.

Author: Fixed. CPI originally stands for "Cloud Provider Interface", but here I would rather use as "Cloud Provider Interface Implementation"

> The Alibaba SLB supports TCP, UDP and HTTPS listeners, an own listener must be configured for each protocol, and those listeners can be assigned to the same SLB instance. 

Reviewer: ```suggestion
The Alibaba SLB supports TCP, UDP and HTTPS listeners. A listener must be configured for each protocol, and then those listeners can be assigned to the same SLB instance. 
```

The number of listeners does not affect SLB pricing.
https://www.alibabacloud.com/help/doc-detail/74809.htm

A user can ask for an internal TCP/UDP Load Balancer via a K8s Service definition that also has the annotation `service.beta.kubernetes.io/alibaba-cloud-loadbalancer-address-type: "intranet"`. Internal SLBs are free.

Summary: Alibaba CPI and SLB seems to support mixed protocol Services, and the pricing is not based on the number of protocols per Service.

#### AWS

> The AWS CPI does not support other protocol in the Service definition than TCP for load balancers. Beside that the AWS CPI looks for annotations on the Service to determine whether TCP, TLS or HTTP(S) listener should be created in the AWS ELB for a configured Service port.

Reviewer: ```suggestion
The AWS CPI does not support mixed protocols in the Service definition since it only allows TCP for load balancers. The AWS CPI looks for annotations on the Service to determine whether TCP, TLS or HTTP(S) listener should be created in the AWS ELB for a configured Service port.
```

AWS Classic LB supports TCP,TLS, and HTTP(S) protocols behind the same IP address. 

AWS Network LB supports TCP/TLS and UDP protocols behind the same IP address. As we can see, UDP cannot be utilized currently, due to the limitation in the AWS CPI.

> If NLB is used the HTTP(S) support would require AWS Application LB, but on the other hand that LB type does not support TCP or UDP, i.e. the usage of TCP+HTTP or UDP+HTTP on the same LB instace behind the same IP address is not possible in AWS.

Reviewer: Might just want to skip the first part and just mention everything after i.e.

Author: Fixed

> From pricing perspective the AWS NLB and the CLB have the following models:

Reviewer: ```suggestion
From a pricing perspective the AWS NLB and the CLB have the following models:
```

https://aws.amazon.com/elasticloadbalancing/pricing/

> Both are rather usage based than "number of protocols" based, though it is true that in case of NLB TCP, UDP and TLS has separated quotas in the pricing units.

Reviewer: ```suggestion
Both are primarily usage based rather than charging based on the number of protocol, however NLBs have separate pricing unit quotas for TCP, UDP and TLS.
```

A user can ask for an internal Load Balancer via a K8s Service definition that also has the annotation `service.beta.kubernetes.io/aws-load-balancer-internal: 0.0.0.0/0`. So far the author could not find any difference in the usage and pricing of those when compared to the external LBs - except the pre-requisite of a private subnet on which the LB can be deployed.

Summary: AWS CPI is the current bottleneck with its TCP-only limitation. As long as it is there the implementation of this feature will have no effect on the AWS bills.

#### Azure

Azure CPI LB documentation: https://github.com/kubernetes-sigs/cloud-provider-azure/blob/master/docs/services/README.md

> The Azure CPI already supports the usage of both UDP and TCP protocols in the same Service definition. It is achieved with the CPI specific annotation `service.beta.kubernetes.io/azure-load-balancer-mixed-protocols`. If this key has value `true` in the Service definition the Azure CPI adds the other protocol value (UDP or TCP) to its internal Service representation, and as a consequence it also manages twice the amount of load balancer rules for the specific frontend. 

Reviewer: ```suggestion
The Azure CPI already supports the usage of both UDP and TCP protocols in the same Service definition. It is achieved with the CPI specific annotation `service.beta.kubernetes.io/azure-load-balancer-mixed-protocols`. If this key has value of `true` in the Service definition, the Azure CPI adds the other protocol value (UDP or TCP) to its internal Service representation. Consequently, it also manages twice the amount of load balancer rules for the specific frontend. 
```

Only TCP and UDP are supported in the current mixed protocol configuration.

The Azure Load Balancer supports only TCP and UDP as protocols. HTTP support would require the usage of the Azure Application Gateway. I.e. HTTP and L4 protocols cannot be used on the same LB instance/IP address

Basic Azure Load Balancers are free.

Pricing of the Standard Azure Load Balancer is based on load-balancing rules and outbound rules.  There is a flat price up to 5 rules, on top of which every new forwarding rule has additional cost. 
https://docs.microsoft.com/en-us/azure/load-balancer/load-balancer-overview#pricing

A user can ask for an internal Azure Load Balancer via a K8s Service definition that also has the annotation `service.beta.kubernetes.io/azure-load-balancer-internal: true`. There is not any limitation on the usage of mixed service protocols with internal LBs in the Azure CPI.

Summary: Azure already has mixed protocol support for TCP and UDP, and it already affects the bills of the users. The implementation of this feature may require some work in Azure CPI.

#### GCE

> The GCE CPI supports both TCP and UDP protocols in Services. Other protocols are not supported.

Reviewer: ```suggestion
The GCE CPI supports only TCP and UDP protocols in Services.
```

> GCE/GKE creates Network Load Balancers based on the K8s Services with type=LoadBalancer. In GCE there are "forwarding rules" that define how the incoming traffic shall be forwarded to the compute instances. A single forwarding rule can support either TCP or UDP but not both. In order to have both TCP and UDP forwarding rules we have to create separate forwarding rule instances for those. Two or more forwarding rules can share the same external IP if

Reviewer: I think the issue/PR has a comment with someone mentioning that this worked for them. We should crosslink comment of the steps that they took as a demonstration of this working.

Author: Fixed. I added a link here to the comment.

- the network load balancer type is External
- the external IP address is not ephemeral but static

HTTP protocol support: there is a different LB type in GCE for HTTP traffic: HTTP(S) LB. I.e. just like in the case of AWS it is not possible to have e.g. TCP+HTTP or UDP+HTTP behind the same LB/IP address.

Forwarding rule pricing is per rule: there is a flat price up to 5 forwarding rule instances, on top of which every new forwarding rule has additional cost. 

https://cloud.google.com/compute/network-pricing#forwarding_rules_charges

> As we can see two different protocols in the same K8s Service definition would result in the creation of 2 forwarding rules in GCP, which has the same fixed price only up to 5 forwarding rule instances, but on top of that every new rule means extra cost.

Reviewer: ```suggestion
[GCP forwarding_rules_charges](https://cloud.google.com/compute/network-pricing#forwarding_rules_charges) suggest that the same K8s Service definition would result in the creation of 2 forwarding rules in GCP. This has the same fixed price up to 5 forwarding rule instances, and each additional rule results in extra cost.
```

A user can ask for an internal TCP/UDP Load Balancer via a K8s Service definition that also has the annotation `cloud.google.com/load-balancer-type: "Internal"`. Forwarding rules are also part of the GCE Internal TCP/UDP Load Balancer architecture, but in case of Internal TCP/UDP Load Balancer it is not supported to define different forwarding rules with different protocols for the same IP address. That is, for Services with type=LoadBalancer and with annotation `cloud.google.com/load-balancer-type: "Internal"` this feature would not be supported.

> Summary: the implementation of this feature can affect the bills of the users. On the other hand, we can see it from these perspectives also:

Reviewer: ```suggestion
Summary: The implementation of this feature can affect the bills of GCP users. However the following perspectives are also observed:
```

> - currently the users cannot have mixed protocols behind the same NLB - it blocks some use cases on GCE

Reviewer: Possibly link to the GCE users in the issue/PR that are affected by this

Author: It is hard to find exact GCE users. I can tell that the one who invented the workaround on GCE is on GCE, but he already has the workaround :) So, it may be not convincing to link that comment here.
